### ISSUE TITLE
[Bug] Gemma2 accuracy issue

### ISSUE THREAD

**Issue by ispobock** [sgl-project/sglang#8614]
### Checklist

- [ ] 1. I have searched related issues but cannot get the expected help.
- [ ] 2. The bug has not been fixed in the latest version.
- [ ] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [ ] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [ ] 5. Please use English, otherwise it will be closed.

### Describe the bug

https://github.com/sgl-project/sglang/actions/runs/16645841906/job/47108486432#step:4:6273 CI failed

**Comment by ispobock** [comment]
Maybe related to https://github.com/sgl-project/sglang/pull/8541.

**Comment by lifuhuang** [comment]
Rerun CI for #8541 is still green: https://github.com/sgl-project/sglang/actions/runs/16638601571/job/47154729476

Let me take a look.

**Comment by ch-wan** [comment]
The main branch can pass the test on my local machine. I think it's an environment issue. How about slightly increasing the threshold?

**Comment by lifuhuang** [comment]
Can't repro on my machine either (single H100).

It's interesting though I can seemingly repro it in CI, I just scheduled two CI runs:
- Latest main (failed): https://github.com/sgl-project/sglang/actions/runs/16659824706/job/47157688213
- My commit yesterday (successful as expected): https://github.com/sgl-project/sglang/actions/runs/16638601571/job/47154729476

It seems that something some changes made between these two commits affected the numerical result.

**Comment by lifuhuang** [comment]
@ch-wan / @ispobock , I am suspecting the issue was introduced by sgl-kernel 0.2.8 as the first commit where the failure can repro (at CI) is this #8550. The commit in main before it was able to pass the failed CI: https://github.com/sgl-project/sglang/actions/runs/16641421280/job/47159722172 

Are you guys aware of any changes since 0.2.7 that might result in changes in numerical behaviors?

If not, probably we should increase the tolerance.

**Comment by ch-wan** [comment]
@lifuhuang After checking sgl-kernel history (https://github.com/sgl-project/sglang/commits/main/sgl-kernel), I feel that these two kernel update may affect numerical behavior

- https://github.com/sgl-project/sglang/commit/0c8dab9e67b1fe0d274a27af03540b2ce5525a37
- https://github.com/sgl-project/sglang/commit/91e3d1542e545c9429b86ad72c526f291ed2b4dc

Other commits are mostly related to MoE, but gemma-2 is a dense model.

cc @zhyncs

**Comment by zhyncs** [comment]
@Fridge003 @yuan-luo

**Comment by Edenzzzz** [comment]
I suspect there might be a bug in the logits_softcap kernel dispatch. In Flashinfer, the kernel dispatch is done during plan ([template param passed in](https://github.com/flashinfer-ai/flashinfer/blob/b9f218fb61fa5c82329fea1506990398ab9da984/csrc/batch_decode_customize_config.jinja#L24) via JIT compile). If you don't pass in logits_softcap during `wrapper.plan`, it will select the kernel that [does not apply any soft capping](https://github.com/flashinfer-ai/flashinfer/blob/b9f218fb61fa5c82329fea1506990398ab9da984/include/flashinfer/attention/variants.cuh#L74).
